runtime-sdk/src/modules/rewards/types.rs

```diff
@@ -0,0 +1,308 @@
+#[serde(deny_unknown_fields)]
+pub struct RewardSchedule {
+    #[serde(rename = "steps")]
+    pub steps: Vec<RewardStep>,
+}
+
+/// Errors emitted during reward schedule validation.
+#[derive(Error, Debug)]
+pub enum RewardScheduleError {
+    #[error("steps not sorted correctly")]
+    StepsNotSorted,
+}
+
+impl RewardSchedule {
+    /// Perform basic reward schedule validation.
+    pub fn validate_basic(&self) -> Result<(), RewardScheduleError> {
+        let mut last_epoch = Default::default();
+        for step in &self.steps {
+            if step.until <= last_epoch {
+                return Err(RewardScheduleError::StepsNotSorted);
+            }
+            last_epoch = step.until;
+        }
+        Ok(())
+    }
+
+    /// Compute the per-entity reward amount for the given epoch based on the schedule.
+    pub fn for_epoch(&self, epoch: beacon::EpochTime) -> token::BaseUnits {
+        for step in &self.steps {
+            if epoch < step.until {
+                return step.amount.clone();
+            }
+        }
+
+        // End of the schedule, default to no rewards.
+        Default::default()
+    }
+}
+
+/// Action that should be taken for a given address when disbursing rewards.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum RewardAction {
+    Reward(u64),
+    NoReward,
+}
+
+impl RewardAction {
+    /// Increment the reward counter associated with the reward.
+    ///
+    /// In case the action is `NoReward` nothing is changed.
+    pub fn increment(&mut self) {
+        match self {
+            RewardAction::Reward(ref mut v) => *v += 1,
+            RewardAction::NoReward => {
+                // Do not change state as the entity has been penalized for the epoch.
+            }
+        }
+    }
+
+    /// Forbids any rewards from accumulating.
+    pub fn forbid(&mut self) {
+        *self = RewardAction::NoReward;
+    }
+
+    /// Value of the reward counter.
+    pub fn value(&self) -> u64 {
+        match self {
+            RewardAction::Reward(v) => *v,
+            RewardAction::NoReward => 0,
+        }
+    }
+}
+
+impl Default for RewardAction {
+    fn default() -> Self {
+        RewardAction::Reward(0)
+    }
+}
+
+/// Rewards for the epoch.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct EpochRewards {
+    #[serde(rename = "pending")]
+    pub pending: BTreeMap<Address, RewardAction>,
+}
+
+impl EpochRewards {
+    /// Returns an iterator over addresses that should be rewarded.
+    pub fn for_disbursement(
+        &self,
+        threshold_numerator: u64,
+        threshold_denominator: u64,
+    ) -> impl Iterator<Item = Address> + '_ {
+        let max_v = self
+            .pending
+            .iter()
+            .fold(0, |acc, (_, action)| std::cmp::max(acc, action.value()));
+        let threshold = (threshold_numerator * max_v) / threshold_denominator;
```

Comment:
this should be fine because we only increment the values by one for anything

Reply:
Added some checks nevertheless.